waveorder/util.py

```diff
@@ -871,52 +872,180 @@ def Dual_variable_ADMM_TV_deconv_2D(AHA, b_vec, rho, lambda_u, lambda_p, itr, ve
+        S0_stack = cp.array(S0_stack.astype('float32'))
+        H_eff = cp.array(H_eff.astype('complex64'))
+        xp = cp
     else:
+        xp = np
+    
+    S0_stack_f = xp.fft.fftn(S0_stack, axes=(-3,-2,-1))
+    N,M,L = S0_stack_f.shape
+    
+    # a "named tuple" representing a point on the L curve
+        # allows for dot notation to access attributes
+    Point_L_curve = namedtuple("Point_L_curve", "reg_x reg data_norm reg_norm f_real_f")
+    
+    def menger_curvature(points):
+        # x is data norm, y is reg norm
+        x, y = xp.zeros(3), xp.zeros(3)
+        for i in range(3):
+            x[i] = points[i].data_norm
+            y[i] = points[i].reg_norm
+        d0 = ((x[1]-x[0])**2+(y[1]-y[0])**2)**(1/2)
+        d1 = ((x[2]-x[1])**2+(y[2]-y[1])**2)**(1/2)
+        d2 = ((x[2]-x[0])**2+(y[2]-y[0])**2)**(1/2)
+
+        signed_area = 0.5*( (x[1]-x[2])*(y[1]-y[0])-(x[1]-x[0])*(y[1]-y[2]) )
+
+        return 4*signed_area / (d0*d1*d2)
+    
+    # evaluate the L curve at a specific lambda
+        # returns a new Point_L_curve() tuple.
+        # this function is the only place where new points are instantiated
+    def eval_L_curve(reg_x):
+        reg_coeff = 10**reg_x
+        H_eff_abs_square = xp.abs(H_eff)**2
+        H_eff_conj = xp.conj(H_eff)
+
+        # FT{f} (f=scattering potential (whose real part is (scaled) phase))
+        f_real_f = S0_stack_f * H_eff_conj / (H_eff_abs_square + reg_coeff)
+        S0_est_stack_f = H_eff * f_real_f # Ax (put estimate through forward model)
+
+        data_norm_eval = xp.log(xp.linalg.norm(S0_est_stack_f - S0_stack_f)**2 /N/M/L)
+        reg_norm_eval = xp.log(xp.linalg.norm(f_real_f)**2 /N/M/L)
+
+        return Point_L_curve(reg_x, reg_coeff, data_norm_eval, reg_norm_eval, f_real_f)
+
+    # creates return value of the whole function
+        # (scaled) phase = real part of inverse FT {scattering potential}
+    def ifft_f_real(f_real_f):
+        f_real = xp.real(xp.fft.ifftn(f_real_f, axes=(-3,-2,-1)))
+        if use_gpu:
+            return cp.asnumpy(f_real)
+        else:
+            return f_real
+    
+    def calc_golden_x(a, b):
+        gs_ratio = (1+xp.sqrt(5))/2
+        return (a*gs_ratio + b) / (1 + gs_ratio)
+    
+    if autotune:
+        # initialize golden section search
+        reg_x_cent = xp.log10(reg_re)  # reg_re becomes middle of search range
+        reg_x = xp.zeros(4)
+        reg_x[0] = reg_x_cent - search_range_auto  # search range = reg_x_cent +/- search_range_auto
+        reg_x[3] = reg_x_cent + search_range_auto
+        reg_x[1] = calc_golden_x(reg_x[0], reg_x[3])
+        reg_x[2] = reg_x[0] + (reg_x[3]-reg_x[1])
+
+        # holds the 4 current points
+        curr_pts = []
+        for i in range(4):
+            curr_pts.append(eval_L_curve(reg_x[i]))
+        
+        opt_list = [] # save optimal point from each iteration
```

Comment:
We might consider saving each optimal points when we want to debug. For real application, if we do not use the history optimal points, we could consider not saving them to save some memory footprint.